CONTRIBUTING.md

```diff
@@ -1,73 +1,110 @@
+  * Code in hpp files is compiled for every cpp file that includes the header. As such, move code to cpp files where
+    possible. This often includes templated classes, where it is sometimes possible to implement their code in cpp
+    files.
+  * Anonymous namespaces are a good way to define local helper methods.
+  * Use forward declarations instead of full header includes wherever possible.
 
-## C++ guidelines
-- Do not use `new` or `malloc`
-- Do not nest namespaces
-- Do not import namespaces (`std::`)
+* Loops
+  * Use range-based `for` loops when possible: `for (const auto& item : items) {...}`.
+  * If you have to use old-style loops, keep in mind that the loop condition is evaluated every time: Instead of
+    `for (auto offset = size_t{0}; offset < something.size(); ++offset)`, the size should be retrieved just once. See
+    also [this document](http://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop).
 
-- When overriding a `virtual` method, avoid repeating `virtual` and always use `override` or `final`
-- Use const (including cbegin() and cend()) whenever possible
-- Use [u]int(8|16|32|64)_t instead of `int, long, uint` etc.
-- Include in this order: header for implementation file, c system, c++ system, other
-- If your templated methods/classes are used outside of your file, they have to be in the header. But if you only use them internally, you should place them in the cpp file.
-- Try to keep the size of your templated methods as small as possible (Thin Template idiom). If only a small part of your method depends on the template parameter, consider moving the rest into a non-templated method. This reduces compile times.
-- Use smart pointers over c-style pointers
-- Use `HYRISE_DEBUG` macro for non-essential checks
-- Be specific: `double a = 3.0;` but `float a = 3.0f;`
-- Use forward declarations whenever possible to reduce compile time
-- We mostly use structs for PODS (plain old data structures). If it has methods, chances are that it is a class.
-- Don't write `this->` if you don't have to
-- Use C++11 for loops when possible: `for (const auto& item : items) {...}`
-- When creating a vector where you know the size beforehand, use `reserve` to avoid unnecessary resizes and allocations
-- Don’t evaluate end() every time through a loop: http://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop
-- Use `if (object) {` over `if (object != nullptr) {` or `if (object.has_value()) {`
+* Data structures
+  * When creating a vector where you know the size beforehand, use `reserve` to avoid unnecessary resizes and
+    allocations.
+  * Hash-based data structures are usually faster than tree-based data structures. Unless you have a reason to use the
+    latter, prefer unordered_(map|set) over map and set.
 
+* Copies
+    * C++ makes it too easy to inadvertently copy objects.
+    * For larger elements (e.g., vectors), pass a (const) reference instead.
```

Comment:
`const`?

Reply:
?

Reply:
Same as above, indicate code. I believe we used `const` over const in other places, too.